### ISSUE TITLE
Is there anyway to rebuild the index without items marked as deleted?

### ISSUE THREAD

**Issue by samehraban** [nmslib/hnswlib#249]
I have an index with a lot of deleted items (currently it contains around 1M vectors but around 700k vectors are marked as deleted). I'm looking for a way to rebuild the index and throw those deleted items away. Since the metric is cosine I don't have original vectors so that I build a new index with remaining vectors and replace the old one with. Any suggestions?

**Comment by yurymalkov** [comment]
Hi @samehraban, the simplest thing to do is to reconstruct the index from scratch. True element deletion/updates is more expensive in terms of complexity compared to insertion.
I am not sure why the cosine similarity might be a blocker. You can retrieve non-deleted elements from the index and build another index with those.

**Comment by samehraban** [comment]
The problem with cosine metric is we don't have access to original vectors, only normalized vectors.

**Comment by yurymalkov** [comment]
@samehraban I am not sure I understand how it blocks your application. 
The index does not care about the norm of the vectors if you are using cosine distance, so it would not affect its performance in any way...

**Comment by samehraban** [comment]
@yurymalkov I need original vectors to build new index but I can't get them via index.

**Comment by yurymalkov** [comment]
Hm. One does not need the exact original vectors to build the index, unless you have an external task that requires it.
I am not sure what am I missing here.

**Comment by samehraban** [comment]
So if I use `get_items(ids)` and build a new index with those vectors, I have almost same index?

**Comment by yurymalkov** [comment]
Yes. Moreover, it should be somewhat faster to search (since the number of elements is less).

**Comment by gkorland** [comment]
@yurymalkov will calling save then load do the trick? or will it save and load the deleted items?

```
save_index(path_to_index)
load_index(path_to_index) 
```

**Comment by yurymalkov** [comment]
@gkorland It will load also the deleted elements. One has to call the construction explicitly only on non-deleted elements to have an index without them.

**Comment by gkorland** [comment]
@yurymalkov or perhaps better not to save those elements on save_index

**Comment by yurymalkov** [comment]
@gkorland If you do not save them you can easily end up with a disconnected graph (=> bad performance). Non-deleted elements have to update their links to maintain the connectivity.
